Clarify that ReducedBeamCharacteristics are relative to reference particle (#585)
BLAST-ImpactX/impactx: docs/source/dataanalysis/dataanalysis.rst

Reviewer: You changed the wording from '**beam** momentum' / '**beam** position' etc to '**particle** momentum'. Is there a particular reason for it? The section title is '**beam** characteristic'. I am confused :)

Reviewer 2: That was me, sorry.  I did this because the word "beam" usually implies a collective property, like "beam energy" or "beam size".  At this point in the documentation, we are referring to the individual particle position and momentum.  (Of course, after the reduced diagnostics are computed, those are properties of the beam--hence the section title.)  Is that too confusing?

Reviewer: I think it's fine! :)

Reduced Beam Characteristics
----------------------------

> ImpactX calculates reduced beam characteristics based on the beam moments during runtime.  These include averaged positions, momenta, beam emittances and Courant-Snyder (Twiss) parameters.
> For computing beam moments (as elsewhere), positions and momenta are given as deviations with respect to the reference particle (see :ref:`Coordinates and Units <theory-coordinates-and-units>`).
> The reduced beam characteristics are calculated before, after, and during each step of the simulation.

Reviewer 3: ```suggestion
ImpactX calculates reduced beam characteristics based on the beam moments during runtime.
These include averaged positions, momenta, beam emittances and Courant-Snyder (Twiss) parameters.
For computing beam moments (as elsewhere), positions and momenta are given as deviations with respect to the reference particle (see :ref:`Coordinates and Units <theory-coordinates-and-units>`).

The reduced beam characteristics are stored with the output of the beam monitor element.
They are also calculated before, after, and during each step of the simulation.
```

----------------------------

> ImpactX calculates reduced beam characteristics like averaged positions, momenta, beam emittances and Courant-Snyder (Twiss) parameters during runtime.

Reviewer: Would it make sense to state in this sentence that this is related to diag.enable, i.e. the '_diag._' section in the input file docs? I remember I was confused by what this section is about, where the reduced diags is specified in the program and actually the whole diagnostics setup. The third sentence where 'diag.slice_step_diagnostic' is mentioned gives a glimpse but describes a detail. (Likewise an explicit pointer to <element_name>.beam_monitor would be helpful in the section above.)

I think what contributes to the confusion is the subtle difference in terminology and a fundamental difference to the other codes (WarpX/HiPACE++). In this section it is called '**reduced** beam diagnostics' but in the inputs description it is called just 'diagnostics' while the holistic 6D diagnostic of the beam can only be done through monitors, which are not specified through the '_diag._' inputs and are not explicitly mentioned in the diagnostics section. While for the other codes with 'diag.' the holistic diagnostic is specified and the in-situ (~reduced) diagnostic appears as an additional feature within the '_diag._' specifications. I guess, this follows the idea, that the reduced diagnostics is the relevant diagnostic but then the changing terminology is confusing :).

Maybe adding a sentence in the 'Parameters: Inputs File/Diagnostics and output' section that clarifies the two options of diagnostics and access points would be helpful? Right now the headline 'Diagnostics and output' kinda infers that '_diag._' specifies THE diagnostic but as much as I understand it mainly relates to the reduced beam diagnostic and 6D phase space openPMD output is generated via the elements. The latter can only be passively grasped when digesting the side notes: 'This option is ignored for the openPMD output elements (remove them from the lattice to disable).' and 'See the beam_monitor element for backend values.'.

Reviewer 3: Thanks Sarah, that is a good point.

Before adding the monitors we had some ad-hoc diagnostics (in ASCII) in ImpacX, where some of the legacy wording here comes from. We do not have a "full" diagnostics outside of monitors in ImpactX, for simplicity (still needs #537).

The "reduced diagnostics" (in situ diagnostics) is now also added to the monitors #584, so that we can technically streamline this even a bit more.

Let's add the cross-section hints you propose for now, to clarify what means what.

Reviewer 2: I made an edit to line 61, so these comments now are invisible under the Files Changed tab.  Just FYI - these are good comments and we don't want to lose them.

> ImpactX calculates reduced beam characteristics like averaged positions, momenta, beam emittances and Courant-Snyder (Twiss) parameters during runtime.

Reviewer 2: ```suggestion
ImpactX calculates reduced beam characteristics based on the beam moments during runtime.  These include averaged positions, momenta, beam emittances and Courant-Snyder (Twiss) parameters.
```

> Averaged positions and momenta are given as deviations with respect to the reference particle (see `Coordinates and Units <theory-coordinates-and-units>`_).

Reviewer: Maybe 'deviation' is a dangerous word here due to its mathematical definition but also plain language usage. Wouldn't just '_Positions and momenta are given with respect to the reference particle_' be less confusing  also because they are treated differently?

Also: it's not only the _averaged_ position and momenta but also minimum and maximum.

Reviewer 3: I think deviation is correctly meant in the mathematical way (difference) here, no? @cemitch99 

There is a small linkage issue here though
```suggestion
Averaged positions and momenta are given as deviations with respect to the reference particle (see :ref:`Coordinates and Units <theory-coordinates-and-units>`).
```

Reviewer 2: I agree with Sarah's comment about the use of "Averaged" here.  I placed an alternative suggestions above.

> Averaged positions and momenta are given as deviations with respect to the reference particle (see `Coordinates and Units <theory-coordinates-and-units>`_).

Reviewer 2: ```suggestion
For computing beam moments (as elsewhere), positions and momenta are given as deviations with respect to the reference particle (see :ref:`Coordinates and Units <theory-coordinates-and-units>`).
```

> These quantities are calculated before, after, and during each step of the simulation.

Reviewer: ```suggestion
The reduced beam characteristics are calculated before, after, and during each step of the simulation.
```

Reviewer: 'These' lost cohesion due to the added sentence above :)

* ``s``, ``ref_beta_gamma``
    Reference particle coordinate ``s`` (unit: meter) and relativistic momentum normalized by the particle mass and the speed of light (unit: dimensionless)
* ``x_mean/min/max``, ``y_mean/min/max``, ``t_mean/min/max``

>     Average / minimum / maximum beam particle displacement with respect to the reference particle in the dimensions of ``x``, ``y`` (transverse coordinates, unit: meter), and ``t`` (normalized time difference :math:`ct`, unit: meter)

Reviewer 2: ```suggestion
    Average / minimum / maximum particle displacement with respect to the reference particle in the dimensions of ``x``, ``y`` (transverse coordinates, unit: meter), and ``t`` (normalized time difference :math:`ct`, unit: meter)
```

* ``sig_x``, ``sig_y``, ``sig_t``
    RMS of the average beam particle positions (unit: meter)
* ``px_mean/min/max``, ``py_mean/min/max``, ``pt_mean/min/max``

>     Average / minimum / maximum beam momentum deviation from the reference momenta and normalized by the longitudinal reference particle momentum (unit: dimensionless, radians for transverse momenta)

Reviewer 2: ```suggestion
    Average / minimum / maximum particle momentum deviation from the reference particle momentum, divided by the magnitude of the reference particle momentum (unit: dimensionless, radians for transverse momenta)
```

Reviewer 2: While we refer to "beam particles" vs. "reference particle", using the word "beam" here could lead to confusion.  I think the meaning is now clear without it.

Author: Good point, thank you Chad! :slightly_smiling_face:

* ``sig_px``, ``sig_py``, ``sig_pt``

>     RMS of the average beam momentum deviations (energy difference for ``pt``) normalized by the longitudinal reference particle momentum (unit: dimensionless)

Reviewer 2: ```suggestion
    Standard deviation of the particle momentum deviations (energy difference for ``pt``) normalized by the magnitude of the reference particle momentum (unit: dimensionless)
```
